feat(@embark/): added feature for contract deployment via the Ledger Hardware Wallet

mul1sh opened the pull request:
This PR will add support for deploying contracts through the ledger hardware wallet. It fixes #825

To test this feature, basically add the following options in the embark's `contracts.js` config.

```
module.exports = {
  development: {
    "deployment": {
      "accounts": [
        {
          "hardwareWallet": "ledger",
          "networkId":"4" , // the id of the ethereum network you are connecting to
          "numAddresses":"1",// number of accounts to derivate/load from device
          "showDeviceConfirmations":"true", //show confirmations in the device when doing transactions,
          "devicePath":"" // you have to set the device path if you're using OSX
        }
      ]
    }
  }
}
```

PascalPrecht commented:
@mul1sh it looks like the "minor changes" commit in this PR (https://github.com/embark-framework/embark/pull/1061/commits/3a0a10665df4bcc3ab9e001d07db2b98825204f7) has actually crucial changes to make your feature in the previous commit work.

Can you confirm?

mul1sh commented:
@PascalPrecht not really, I have initially expanded the contracts config object `this.config.contractsConfig.deployment` to include support for the options needed to create the ledger provider. 

But then i realised this was unnecessary as the same options could easily be passed in the accounts array as was actually required in the i[ssue](https://github.com/embark-framework/embark/issues/825#issue-359888507) so thats why removed those changes and this feature should still work without them

mul1sh commented:
@andremedeiros you can also test with your ledger device and see how it works 🙂

PascalPrecht commented:
Hey @mul1sh CI still seems to be failing. I brought my HW today to test this branch.

Let me know when it is in a reviewable state :)

mul1sh commented:
@PascalPrecht will do, looking into why the tests are failing now 🙂

mul1sh commented:
@PascalPrecht had a chance to test this out yet? thanks 🙂

jrainville commented on src/lib/modules/blockchain_connector/provider.js:
```diff
@@ -16,9 +20,31 @@ class Provider {
 
   startWeb3Provider(callback) {
     const self = this;
+    const accountsConfig = self.accountsConfig[0];
```
Can you put this variable inside the `else if (this.type === 'ledger') {` as it is the only place where it is used.

jrainville replied:
Also, I see you only check the first account of the config. Does that mean that if the ledger account is not the first, it's completely ignored?

jrainville commented on src/lib/modules/blockchain_connector/provider.js:
```diff
@@ -16,9 +20,31 @@ class Provider {
     if (this.type === 'rpc') {
       self.provider = new this.web3.providers.HttpProvider(self.web3Endpoint);
+    } else if (this.type === 'ledger') {
+        if (accountsConfig.HardwareWallet === 'Ledger') {
```
`accountsConfig` can be `undefined`, so you need to check it exists first.

jrainville commented on src/lib/modules/blockchain_connector/provider.js:
```diff
@@ -16,9 +20,31 @@ class Provider {
+
+          engine.addProvider(ledger); 
+          engine.addProvider(new FetchSubprovider({ rpcUrl }));
+          engine.start();
```
Could you put the `start` after the event listeners below, just in case the engine fails immediately.

jrainville commented on src/lib/modules/blockchain_connector/provider.js:
```diff
@@ -16,9 +20,31 @@ class Provider {
+          self.provider.on('error', (err) => self.logger.error('Error while connecting to the ledger wallet ', (err.message || err.stack) || err));
+          self.provider.on('end', () => self.logger.error('Connection to wallet ended'));
+
+        }
```
I cannot quite comment on it because it is too low, but below, we call `AccountParser.parseAccountsConfig`, which checks the `accountsConfig`. Since you added a new account type (ledger) and it's not supported by the AccountParser (since it's handled by the provider), it will give a warning saying `Unsupported account configuration`. You should add a condition in `AccountParser.getAccountgetAccount()` that just ignores the `ledger` account.

mul1sh replied:
@jrainville good catch on this 🙂 i had actually disabled the AccountParser part when testing on my end because i was getting the `Unsupported account configuration: ` msg. I forgot i needed to do this check when i re-enabled it in order to push.

PascalPrecht commented:
@mul1sh I just tested this locally and running

```
$ embark run develop --nodashboard --nobrowser
```

Actually seems to be stuck as it doesn't find a blockchain node. Can you confirm this?

mul1sh commented:
@PascalPrecht ok, did you enter the hardware address and specify the `config` options as outlined above?

PascalPrecht commented:
I used the exact config above, but I'm honestly not sure if the `...` part indicates "enter your specific address here". Might that be the problem?

mul1sh commented:
@PascalPrecht  ahh ok. I've updated the code as per @jrainville comment's, so now please change the config to follow the above updated format.

Also note the `walletAddress` is the address to the device connected via usb as returned by node-hid (https://github.com/node-hid/node-hid#list-all-hid-devices-connected). Maybe i could add a simple util to help people figure out this address, because this also confused me a great deal on my end

mul1sh commented:
@PascalPrecht i've fixed a merge conflict and some failing tests, please feel free to test again, looking forward to your feedback 🙂

jrainville commented on src/lib/modules/blockchain_connector/provider.js:
```diff
@@ -16,9 +20,34 @@ class Provider {
     if (this.type === 'rpc') {
       self.provider = new this.web3.providers.HttpProvider(self.web3Endpoint);
+    } else if (this.type === 'ledger') {
+        self.accountsConfig.forEach((accountConfig) => {
```
We should print an error if there is no `hardwareWallet: 'ledger'` inside the accountsConfig, since we have no way to connect to it.

Also, what should we do if a user puts two accounts with a wallet config? We should at least warn and pick only the first one.

mul1sh replied:
@jrainville done, thanks again for the great feedback 🙂

PascalPrecht commented:
@mul1sh I must be doing something horribly wrong. Even with the exact config given above (with my ledger's path), Embark doesn't successfully spin up all the processes in your branch.

Can you confirm this?

jrainville commented on src/lib/modules/blockchain_connector/provider.js:
```diff
@@ -16,9 +20,48 @@ class Provider {
+          }
+        });
+
+        if (hardwareWalletConfigs.length > 0) {
```
To make it clearer and cleaner, you could do:
```
if (!hardwareWalletConfigs.length) {
  // TODO when we have documentation for it, add a link in the error message
  return callback(__('No ledger wallet config found, please add it'));
}
```
Notice how I return and callback immediately. That's because, even if we went through with the initialization, it would bug because thee is no account.
Also, that way, you don't need an `else`

jrainville commented on src/lib/modules/blockchain_connector/provider.js:
```diff
@@ -16,9 +20,48 @@ class Provider {
+
+        if (hardwareWalletConfigs.length > 0) {
+          if (hardwareWalletConfigs.length > 1) {
+            self.logger.warn('Mulitple ledger wallet configs found, so selecting the first config only.');
```
```suggestion
            self.logger.warn(__('Multiple ledger wallet configs found. Selecting the first config only.'));
```
Fixed a type and also added `__` as it handles translations.

mul1sh commented:
@jrainville awesome thanks for the feedback, really appreciate it 🙂

PascalPrecht commented:
@mul1sh just gave this one a spin again. If your latest changes in the branch I'm getting:

```
/Users/pascalprecht/projects/embark/embark/src/lib/modules/blockchain_connector/provider.js:105                                                                                  
      const realSend = self.provider.send.bind(self.provider);
                                     ^
TypeError: Cannot read property 'send' of undefined
    at send (/Users/pascalprecht/projects/embark/embark/src/lib/modules/blockchain_connector/provider.js:105:38)                                                                 
    at Object._fireError (/Users/pascalprecht/projects/embark/embark/node_modules/web3/node_modules/web3-utils/src/index.js:60:9)                                                
    at sendTxCallback (/Users/pascalprecht/projects/embark/embark/node_modules/web3/node_modules/web3-core-method/src/index.js:493:30)                                           
    at RequestManager.send (/Users/pascalprecht/projects/embark/embark/node_modules/web3/node_modules/web3-core-requestmanager/src/index.js:128:16)                              
    at sendRequest (/Users/pascalprecht/projects/embark/embark/node_modules/web3/node_modules/web3-core-method/src/index.js:562:42)                                              
    at Eth.send [as getAccounts] (/Users/pascalprecht/projects/embark/embark/node_modules/web3/node_modules/web3-core-method/src/index.js:583:13)                                
    at Provider.getAccounts [as startWeb3Provider] (/Users/pascalprecht/projects/embark/embark/src/lib/modules/blockchain_connector/provider.js:86:19)                           
    at Function.startWeb3Provider (/Users/pascalprecht/projects/embark/embark/src/lib/modules/blockchain_connector/index.js:164:21)                                              
    at apply (/Users/pascalprecht/projects/embark/embark/src/lib/core/processes/processManager.js:84:14)                                                                         
    at EventEmitter.callback (/Users/pascalprecht/projects/embark/embark/src/lib/modules/blockchain_process/index.js:115:7)                                                      
    at Object.onceWrapper (events.js:273:13)
    at EventEmitter.emit (events.js:182:13)
    at Object.emit [as callback] (/Users/pascalprecht/projects/embark/embark/src/lib/modules/blockchain_process/blockchainProcessLauncher.js:46:19)                              
    at callback (/Users/pascalprecht/projects/embark/embark/src/lib/core/processes/processLauncher.js:123:24)                                                                    
    at Array.forEach (<anonymous>)
    at ProcessLauncher.forEach [as _checkSubscriptions] (/Users/pascalprecht/projects/embark/embark/src/lib/core/processes/processLauncher.js:122:31)                            
    at ChildProcess._checkSubscriptions (/Users/pascalprecht/projects/embark/embark/src/lib/core/processes/processLauncher.js:57:12)                                             
    at ChildProcess.emit (events.js:182:13)
    at emit (internal/child_process.js:811:12)
    at process._tickCallback (internal/process/next_tick.js:63:19)
```

Can you reproduce this?

PascalPrecht commented:
@mul1sh tried it again. Rebased your branch on top of latest `master`. Set the config, disabled browser support, and now it's hanging at "Blockchain node is ready".

PascalPrecht commented:
@mul1sh please try this branch here https://github.com/embark-framework/embark/tree/feat/ledger-support and tell me you're experiencing the same.

PascalPrecht commented:
For transparency, @mul1sh and I just had another call to walk through PR and to demonstrate that it's working on his machine (as I can't get a working state reproduced). Turns out that several things are off in this effort and probably haven't been communicated clear enough from our side:

- So far, this PR has never really been tested with the `default` contracts.js configuration, which connects to the locally installed blockchain node (geth or parity atm).
- This is also why the recommended config above has
  a. A `networkId`, which doesn't belong into an `accounts` config
  b. An `rpcUrl` to connect to infura. Usually if one wants to connect to infura, this is done via `host`, `port`, `protocol` and `type` properties of an environment
- The first thing that needs to happen now is that this PR works with the default configuration, so that the ledger wallet can be used to connect to the local node
- Once that is sorted out, it should work with a dedicated infura configuration out of the box.
- The reason this didn't came up earlier was because @mul1sh didn't actually have a `geth` client installed and wasn't aware that Embark spins up a node for him in the first place

I will update the original issue to reflect the requirements.

@mul1sh let me know if I'm missing anything here.

mul1sh commented:
> For transparency, @mul1sh and I just had another call to walk through PR and to demonstrate that it's working on his machine (as I can't get a working state reproduced). Turns out that several things are off in this effort and probably haven't been communicated clear enough from our side:
> 
> * So far, this PR has never really been tested with the `default` contracts.js configuration, which connects to the locally installed blockchain node (geth or parity atm).
> * This is also why the recommended config above has
>   a. A `networkId`, which doesn't belong into an `accounts` config
>   b. An `rpcUrl` to connect to infura. Usually if one wants to connect to infura, this is done via `host`, `port`, `protocol` and `type` properties of an environment
> * The first thing that needs to happen now is that this PR works with the default configuration, so that the ledger wallet can be used to connect to the local node
> * Once that is sorted out, it should work with a dedicated infura configuration out of the box.
> * The reason this didn't came up earlier was because @mul1sh didn't actually have a `geth` client installed and wasn't aware that Embark spins up a node for him in the first place
> 
> I will update the original issue to reflect the requirements.
> 
> @mul1sh let me know if I'm missing anything here.

sounds great @PascalPrecht and yes true i've never tested with the `default` config just infura, with the above config. Anyway installing geth now and testing the default config with it and then also i'll test with infura again. Thanks again for your patience and feedback

mul1sh commented:
@PascalPrecht i've modified the commit according and also modified the config, so it should now look like 
```
 development: {
    "deployment": {
        "accounts": [
          {
            "hardwareWallet": "ledger",
            "networkId":"4" , // the id of the network you are connecting to with the ethereum node
            "numAddresses":"1",// number of accounts to derivate/load from device
            "showDeviceConfirmations":"true" //show confirmations when doing 
          }
        ]
    }
}
```

I've also tested it thoroughly in my end and so far no issues, thanks again for the feedback, really appreciate it 🙂

PascalPrecht commented:
Hi @mul1sh I just tried again. Unfortunately, it looks like Embark doesn't seem to be able to read accounts from the HW in your branch:

```
❯ ../../bin/embark run --nodashboard --nobrowser
pid 25962 listening on /Users/pascalprecht/projects/embark/embark/templates/demo/.embark/embark.ipc
========================
Welcome to Embark 4.0.0-alpha.1
========================
ready to watch file changes
Starting Blockchain node in another process
IPFS node not found, attempting to start own node
Starting ipfs process
Webserver available at http://localhost:8000
Blockchain node is ready
Connecting to the ledger device, please make sure "browser support" is set to "no" in the device
Error while getting the node's accounts. cannot open device with path IOService:/AppleACPIPlatformExpert/PCI0@0/AppleACPIPCI/XHC1@14/XHC1@14000000/HS02@14100000/USB2.0 Hub@14100000/AppleUSB20Hub@14100000/AppleUSB20HubPort@14110000/Nano S@14110000/Nano S@0/IOUSBHostHIDDevice@14110000,0
Cannot use nodeAccounts in this context
Connecting to the ledger device, please make sure "browser support" is set to "no" in the device
undefined
Ethereum node detected..
loading solc compiler..
compiling solidity contracts...
Connecting to the ledger device, please make sure "browser support" is set to "no" in the device
undefined
```

Notice that I did set browser support to "no"

PascalPrecht commented:
Still breaking with latest changes:

```
Welcome to Embark 4.0.0-alpha.1
========================
ready to watch file changes
Starting Blockchain node in another process
IPFS node not found, attempting to start own node
Starting ipfs process
Webserver available at http://localhost:8000
Blockchain node is ready
Connecting to the ledger device, please make sure "browser support" is set to "no" in the device
(node:51658) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'length' of undefined
    at Provider.length [as fundAccounts] (/Users/pascalprecht/projects/embark/embark/src/lib/modules/blockchain_connector/provider.js:242:24)
    at fundAccounts (/Users/pascalprecht/projects/embark/embark/src/lib/modules/blockchain_connector/index.js:177:23)
    at Provider.callback [as startWeb3Provider] (/Users/pascalprecht/projects/embark/embark/src/lib/modules/blockchain_connector/provider.js:70:16)
    at process._tickCallback (internal/process/next_tick.js:68:7)
(node:51658) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)
(node:51658) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
Error: Provider not set or invalid
    at Object.InvalidProvider (/Users/pascalprecht/projects/embark/embark/node_modules/web3-eth/node_modules/web3-core-helpers/src/errors.js:38:16)
    at RequestManager.send (/Users/pascalprecht/projects/embark/embark/node_modules/web3-eth/node_modules/web3-core-requestmanager/src/index.js:128:32)
    at sendRequest (/Users/pascalprecht/projects/embark/embark/node_modules/web3-eth/node_modules/web3-core-method/src/index.js:562:42)
    at Net.send [as getId] (/Users/pascalprecht/projects/embark/embark/node_modules/web3-eth/node_modules/web3-core-method/src/index.js:583:13)
    at BlockchainConnector.getId [as getNetworkId] (/Users/pascalprecht/projects/embark/embark/src/lib/modules/blockchain_connector/index.js:625:30)
    at getNetworkId (/Users/pascalprecht/projects/embark/embark/src/lib/modules/blockchain_connector/index.js:165:14)
    at Provider.callback [as startWeb3Provider] (/Users/pascalprecht/projects/embark/embark/src/lib/modules/blockchain_connector/provider.js:70:16)
    at process._tickCallback (internal/process/next_tick.js:68:7)
```

PascalPrecht commented:
After several updates, this is still not working:

```
❯ ../../bin/embark run --nodashboard --nobrowser
./chains.json file not found, creating it...
pid 22235 listening on /Users/pascalprecht/projects/embark/embark/templates/demo/.embark/embark.ipc
========================
Welcome to Embark 4.0.0-alpha.1
========================
ready to watch file changes
Starting Blockchain node in another process
IPFS node not found, attempting to start own node
Starting ipfs process
Webserver available at http://localhost:8000
Blockchain node is ready
Testing connection to the ledger device, please make sure "browser support" is set to "no" in the device
Ledger device connection error. Please make sure the device is connected properly and that no other app is using it
(node:22235) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'length' of undefined
    at Provider.length [as fundAccounts] (/Users/pascalprecht/projects/embark/embark/src/lib/modules/blockchain_connector/provider.js:249:24)
    at fundAccounts (/Users/pascalprecht/projects/embark/embark/src/lib/modules/blockchain_connector/index.js:177:23)
    at Provider.callback [as startWeb3Provider] (/Users/pascalprecht/projects/embark/embark/src/lib/modules/blockchain_connector/provider.js:70:16)
    at process._tickCallback (internal/process/next_tick.js:68:7)
(node:22235) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)
(node:22235) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
Error: Provider not set or invalid
    at Object.InvalidProvider (/Users/pascalprecht/projects/embark/embark/node_modules/web3-core-helpers/src/errors.js:38:16)
    at RequestManager.send (/Users/pascalprecht/projects/embark/embark/node_modules/web3-core-requestmanager/src/index.js:128:32)
    at sendRequest (/Users/pascalprecht/projects/embark/embark/node_modules/web3-core-method/src/index.js:562:42)
    at Net.send [as getId] (/Users/pascalprecht/projects/embark/embark/node_modules/web3-core-method/src/index.js:583:13)
    at BlockchainConnector.getId [as getNetworkId] (/Users/pascalprecht/projects/embark/embark/src/lib/modules/blockchain_connector/index.js:625:30)
    at getNetworkId (/Users/pascalprecht/projects/embark/embark/src/lib/modules/blockchain_connector/index.js:165:14)
    at Provider.callback [as startWeb3Provider] (/Users/pascalprecht/projects/embark/embark/src/lib/modules/blockchain_connector/provider.js:70:16)
    at process._tickCallback (internal/process/next_tick.js:68:7)
```

Exact same error as previously commented.

Config:

```
module.exports = {
  // default applies to all environments
  default: {
    // Blockchain node to deploy the contracts
    deployment: {
      host: "localhost", // Host of the blockchain node
      port: 8546, // Port of the blockchain node
      type: "ws" // Type of connection (ws or rpc),
      // Accounts to use instead of the default account to populate your wallet
      /*,accounts: [
        {
          privateKey: "your_private_key",
          balance: "5 ether"  // You can set the balance of the account in the dev environment
                              // Balances are in Wei, but you can specify the unit with its name
        },
        {
          privateKeyFile: "path/to/file", // Either a keystore or a list of keys, separated by , or ;
          password: "passwordForTheKeystore" // Needed to decrypt the keystore file
        },
        {
          mnemonic: "12 word mnemonic",
          addressIndex: "0", // Optionnal. The index to start getting the address
          numAddresses: "1", // Optionnal. The number of addresses to get
          hdpath: "m/44'/60'/0'/0/" // Optionnal. HD derivation path
        }
      ]*/
    },
    // order of connections the dapp should connect to
    dappConnection: [
      "$WEB3",  // uses pre existing web3 object if available (e.g in Mist)
      "ws://localhost:8546",
      "http://localhost:8545"
    ],

    gas: "auto",

    // Strategy for the deployment of the contracts:
    // - implicit will try to deploy all the contracts located inside the contracts directory
    //            or the directory configured for the location of the contracts. This is default one
    //            when not specified
    // - explicit will only attempt to deploy the contracts that are explicity specified inside the
    //            contracts section.
    //strategy: 'implicit',

    contracts: {
      SimpleStorage: {
        fromIndex: 0,
        args: [100]
      }
    }
  },

  // default environment, merges with the settings in default
  // assumed to be the intended environment by `embark run`
  development: {
    dappConnection: [
      "ws://localhost:8546",
      "http://localhost:8545",
      "$WEB3"  // uses pre existing web3 object if available (e.g in Mist)
    ],
    "deployment": {
      "accounts": [
        {
          "hardwareWallet": "ledger",
          "networkId":"4" , // the id of the ethereum network you are connecting to
          "numAddresses":"1",// number of accounts to derivate/load from device
          "showDeviceConfirmations":"true" //show confirmations in the device when doing transactions
        }
      ]
    }
  },

  // merges with the settings in default
  // used with "embark run privatenet"
  privatenet: {
  },

  // merges with the settings in default
  // used with "embark run testnet"
  testnet: {
  },

  // merges with the settings in default
  // used with "embark run livenet"
  livenet: {
  },

  // you can name an environment with specific settings and then specify with
  // "embark run custom_name" or "embark blockchain custom_name"
  //custom_name: {
  //}
};
```

PascalPrecht commented:
Alright, chances are this is related to the OS I'm using (Mac OS High Sierra) as there's a bug with that in LedgerJS: https://github.com/LedgerHQ/ledgerjs/issues/213

@jrainville as you're on windows, would you mind giving this PR a shot and test whether it works?
